source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,344 @@ namespace EnvoyDefault {
+  //                / DIGIT / ALPHA
+  //                ; any VCHAR, except delimiters
+  //
+  const auto& key_string_view = name.getStringView();
+  bool allow_underscores = !config_.reject_headers_with_underscores();
+  // This header name is initially invalid if the name is empty.
+  if (key_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  bool is_valid = true;
+  char c = '\0';
+
+  for (std::size_t i{0}; i < key_string_view.size() && is_valid; ++i) {
+    c = key_string_view.at(i);
+    is_valid = testChar(kGenericHeaderNameCharTable, c) && (c != '_' || allow_underscores);
+  }
+
+  if (!is_valid) {
+    auto details = c == '_' ? UhvResponseCodeDetail::get().InvalidUnderscore
+                            : UhvResponseCodeDetail::get().InvalidCharacters;
+    return {RejectAction::Reject, details};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateRequestHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateGenericHeaderValue(const HeaderString& value) {
+  //
+  // Verify that the header value is valid.
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-value    = *( field-content / obs-fold )
+  // field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
+  // field-vchar    = VCHAR / obs-text
+  // obs-text       = %x80-FF
+  //
+  // VCHAR          =  %x21-7E
+  //                   ; visible (printing) characters
+  //
+  const auto& value_string_view = value.getStringView();
+  bool is_valid = true;
+
+  for (std::size_t i{0}; i < value_string_view.size() && is_valid; ++i) {
+    is_valid = testChar(kGenericHeaderValueCharTable, value_string_view.at(i));
+  }
+
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidCharacters};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateResponseHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateContentLengthHeader(const HeaderString& value) {
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2:
+  //
+  // Content-Length = 1*DIGIT
+  //
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
+  auto buffer_start = value_string_view.data();
+  auto buffer_end = buffer_start + value_string_view.size();
+
+  std::uint32_t int_value{};
+  auto result = std::from_chars(buffer_start, buffer_end, int_value);
+  if (result.ec == std::errc::invalid_argument || result.ptr != buffer_end) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
```

Comment:
will this allow a length which would overflow an [u]int64?    I think we should also check length here to make sure we cap it.

Reply:
[`std::from_chars`](https://en.cppreference.com/w/cpp/utility/from_chars) will detect overflows and return an error:

> If the pattern was matched, but the parsed value is not in the range representable by the type of value, returns value of type from_chars_result such that ec equals [std::errc::result_out_of_range](https://en.cppreference.com/w/cpp/error/errc) and ptr points at the first character not matching the pattern. value is unmodified.